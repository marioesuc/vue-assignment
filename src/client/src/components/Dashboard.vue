<template>
  <div id="dashboard">
    <div class="row">
      <Map
        :latitude="data.gps.latitude"
        :longitude="data.gps.longitude"
      />
      <div class="column">
        <div class="row">
          <div class="column">
          <CircularBar
            :currentValue="data.speed"
            :maxValue="maxSpeed"
            :unit="'km/h'"
          />
          </div>
          <div class="column">
          <CircularBar
            :currentValue="data.soc"
            :maxValue="maxCharge"
            :unit="'%'"
          />
          </div>
        </div>
        <div class="row">
          <div class="column">
          <LargeLabel
            :currentValue="data.energy"
            :unit="'kW'"
          />
          </div>
          <div class="column">
            <LargeLabel
              :currentValue="data.odo"
              :unit="'km'"
            />
          </div>
        </div>
      </div>
    </div>
    <div class="row">
      <LinearChart
        :speed="data.speed"
        :soc="data.soc"
        :time="data.time"
      />
    </div>
  </div>
</template>

<script>
import Map from "./Map";
import CircularBar from "./CircularBar";
import LinearChart from "./LinearChart";
import LargeLabel from "./LargeLabel";
import { MAX_SPEED, MAX_CHARGE } from "@/common/Constants";

export default {
  name: "Dashboard",
  components: {
    Map,
    CircularBar,
    LinearChart,
    LargeLabel,
  },
  props: {
    data: Object
  },
  data() {
    return {
      maxSpeed: MAX_SPEED,
      maxCharge: MAX_CHARGE
    }
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
#dashboard {
  width: 1350px;
  background-color: white;
  margin: 30px auto;
  padding: 20px;
  box-shadow: 0px 5px 15px 0px rgba(0,0,0,0.75);
}
.row {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: 100%;
  justify-content: center;
}
.column {
  display: flex;
  flex-direction: column;
  flex: 1;
  justify-content: center;
  align-items: center;
}
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}
</style>
