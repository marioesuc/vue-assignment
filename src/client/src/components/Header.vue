<template>
  <div id="header">
    <img id="image" :src="logoSrc" />
    <div id="separator"></div>
    <span id="headerText">{{ title }}</span>
  </div>
</template>

<script>
export default {
  props: {
    title: String,
    logoSrc: String
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import "@/common/Colors";

#header {
  width: 100%;
  background-color: $header-background;
  box-shadow: 0px 5px 15px 0px rgba(0, 0, 0, 0.75);
  height: 90px;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: row;
}
#image {
  width: 220px;
  padding: 15px;
}
#separator {
  height: 50px;
  width: 1px;
  background-color: $header-separator;
}
#headerText {
  color: $header-text;
  font-size: 35px;
  font-weight: bold;
  text-align: center;
  padding: 10px;
}

@media only screen and (max-width: 800px) {
  #header {
    height: 75px;
  }
  #image {
    width: 180px;
  }
  #separator {
    height: 50px;
  }
  #headerText {
    font-size: 28px;
  }
}
</style>
