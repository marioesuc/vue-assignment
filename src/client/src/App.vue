<template>
  <div id="app">
    <Header title="Dashboard" :logoSrc="logoSrc" />
    <Dashboard :data="data" />
  </div>
</template>

<script>
import Dashboard from "./components/Dashboard.vue";
import Header from "./components/Header.vue";

export default {
  name: "App",
  components: {
    Dashboard,
    Header
  },
  data() {
    return {
      data: {
        time: 0,
        energy: 0,
        gps: {
          latitude: 0,
          longitude: 0
        },
        odo: 0,
        speed: 0,
        soc: 0
      },
      logoSrc: require("@/assets/logo.png")
    };
  },
  mounted() {
    // Perform connection to WebSocket
    const url = "ws://localhost:3000";
    this.connectToWebsocket(url, this.onConnectToWebsocket);
  },
  methods: {
    connectToWebsocket(url, onConnectToWebsocket) {
      const connection = new WebSocket(url);

      connection.onopen = () => {
        console.log("Connection to WebSocket Succesful");
      };
      connection.onmessage = ({ data }) => {
        onConnectToWebsocket(data);
      };
      connection.onerror = error => {
        console.log(`Error on WebSocket connection: ${error}`);
      };
    },
    onConnectToWebsocket(data) {
      // Parse the data sent by the WebSocket
      const parsedData = JSON.parse(data);
      // Split the string of coordinates in an array of two elements
      const parsedCoordinates = parsedData.gps.split("|");
      // Save into variables all the data
      this.data.time = parseInt(parsedData.time);
      this.data.energy = parseFloat(parsedData.energy);
      this.data.gps.latitude = parseFloat(parsedCoordinates[0]);
      this.data.gps.longitude = parseFloat(parsedCoordinates[1]);
      this.data.odo = parseFloat(parsedData.odo);
      this.data.speed = parseInt(parsedData.speed) || 0;
      this.data.soc = parsedData.soc;
    }
  }
};
</script>

<style>
html,
body,
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  background: #7ca363;
  background: linear-gradient(
    0deg,
    rgba(124, 163, 99, 1) 0%,
    rgba(160, 228, 116, 1) 100%
  );
  background-attachment: fixed;
  height: 100vh;
  margin: 0;
}
</style>
