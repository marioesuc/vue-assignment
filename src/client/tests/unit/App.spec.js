import { shallowMount } from "@vue/test-utils";
import App from "@/App.vue";
import Header from "@/components/Header.vue";
import Dashboard from "@/components/Dashboard.vue";

describe("App.vue", () => {
  it("renders correctly", () => {
    // Stub dynamically loaded components replacing them with their real components
    const stubs = {
      Header,
      Dashboard,
      // Stub LinearChart to fix failing tests because of canvas
      LinearChart: "<span></span>"
    };

    const wrapper = shallowMount(App, { stubs });
    expect(wrapper.find(Header).exists()).toBeTruthy();
    expect(wrapper.find(Dashboard).exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
