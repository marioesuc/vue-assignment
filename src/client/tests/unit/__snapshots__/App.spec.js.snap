// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue renders correctly 1`] = `
<div id="app">
  <div id="header"><img id="image" src="">
    <div id="separator"></div> <span id="headerText">Dashboard</span>
  </div>
  <div id="dashboard">
    <div class="row">
      <div class="column">
        <!---->
      </div>
      <div class="column">
        <div class="row">
          <div class="column">
            <!---->
          </div>
          <div class="column">
            <!---->
          </div>
        </div>
        <div class="row">
          <div class="column">
            <!---->
          </div>
          <div class="column">
            <!---->
          </div>
        </div>
      </div>
    </div>
    <div class="row"><span data="[object Object]" options="[object Object]" intervalenabled="" refreshrate="250"></span></div>
  </div>
</div>
`;
